### DISCUSSION TITLE
[자료공유] 서버와 클라이언트? 프론트엔드, 백엔드?

### DISCUSSION THREAD

**Discussion by dev-LeoKang** [4Moyede/HexatonClass-01#178] [04_자료공유]
안녕하세요 강형구 멘토입니다
여러분들의 데이터베이스 서버 활용을 위해 제가 이해하고 있는 **서버와 클라이언트가 무엇인지** 적어보려고 합니다

우리가 구글에서 검색을 하는 과정을 예시로 들어보겠습니다.

## Server, Client

크롬, 사파리, 네이버 웨일과 같은 웹 브라우저는 우리(Client)가 어떤 http 주소를 입력하면 그 주소에 해당하는 컴퓨터(Server)에 특정한 Resquest을 보내주고 돌아온 Response인 html 코드를 우리가 보기 편하게 해주는 프로그램입니다.

![image](https://user-images.githubusercontent.com/38308646/107139599-bede9580-695f-11eb-887d-82133702e49a.png)

어떤 Request를 보낼지는 URL을 통해 표현할 수 있습니다.

![image](https://user-images.githubusercontent.com/38308646/107139639-2eed1b80-6960-11eb-9346-e9d11c1a8c87.png)

* Protocol : 어떤 방식으로
* Host : 어디의
* Port : 어떤 문으로, 아무것도 적지 않으면 기본값인 80으로
* Resource Path : 어디에 있는 파일에게 요청을 보냅니다
* Query : 아 그리고 이런 변수에 이런 값을 담아서 보냅니다

Google 에 'kist 유혁준'을 검색한 URL 을 예시로 들어보면,

https://www.google.com/search?q=kist+유혁준 이라는 URL은,
"https 방식으로 www.google.com 이라는 host에게 80번 문으로 보내는데, search라는 파일에게 q라는 변수에 'kist+유혁준'을 담아서 보냅니다." 가 됩니다.

여기서 protocol은, client와 server가 주고받는, 또는 프론트엔드 백엔드가 서로 주고받는 request와 response의 형태를 말하며 다음 블로그 포스팅을 참고하시면 됩니다.
https://velog.io/@teddybearjung/HTTP-구조-및-핵심-요소

## Frontend, Backend

여기서 서버에서 일어나는 일을 세분화 시켜보면,

![image](https://user-images.githubusercontent.com/38308646/107140329-92794800-6964-11eb-991f-01cac5205cc4.png)

라고 표현할 수 있습니다.

저는 **프론트엔드**는 어떻게 화면을 구성해서 정보를 보여주는 것이 효과적일까를 고민하는 것이고, **백엔드**는 요청을 처리하려면 어떻게 데이터베이스 및 다른 서비스(AWS, 여러분들이 짜고 있는 함수)를 연결해줘야하는가를 고민하는 것이라고 생각합니다. 

Django, Node.js 등등의 프레임워크를 이용하여서 프론트엔드만 혹은 백엔드만 구축할 수 도 있으며 두 가지의 역할을 함께 하도록 할 수 도 있습니다. 하지만, 각각의 프레임워크들이 가지는 강점들이 있으므로 그에 맞는 프레임워크를 사용할 때에 더 빠르고 기능성이 있는 서버를 만들 수 있습니다.

## 적용

이 과정을 멘티님들의 프로젝트에 적용시켜 보자면,

* 구름IDE에 데이터베이스를 구축하고 데이터베이스에 쿼리문을 보내고 결과를 받을 방법 모색
* 본인의 컴퓨터에 데이터베이스, 프론트, 백엔드 구축하고 클라이언트가 접속할 수 있는 방법 모색

입니다. 찾아봐야하는 검색의 키워드는, **python과 데이터베이스 연동, 포트포워딩, DNS** 정도가 있을 것 같네요.
이해가 잘 되지 않거나 구글링으로 답을 찾지 못한 경우 언제든지 issue에 등록해주시면 될 것 같습니다.

화이팅ㅎ

**Comment by dlrudgnsdlek** [comment]
오 감사합니다!
